sympy/physics/quantum/grover.py

```diff
@@ -324,3 +324,111 @@ def apply_grover(oracle, nqubits, iterations=None):
+    min_pic : int
+        Minimum number of invers images that are mapped to 1
+    max_pic : int
+        Maximum number of invers images that are mapped to 1
+    q_type : OracleGate
+        Type of the Qubits that the oracle should be applied on.
+        Can be 'bin' for binary (Qubit()) or 'int' for integer (IntQubit()).
+
+    Returns
+    =======
+
+    OracleGate : random OracleGate under the given parameter
+
+    Examples
+    ========
+
+    Generate random OracleGate that outputs 1 for 2-4 inputs::
+
+        >>> from sympy.physics.quantum.grover import random_oracle
+        >>> oracle = random_oracle(4, min_img=2, max_img=4, q_type="bin")
+
+    """
+    if q_type != 'bin' and q_type != 'int':
+        raise QuantumError("q_type must be 'int' or 'bin'")
+
+    if min_img < 1 or max_img < 1:
+        raise QuantumError("min_pic, max_pic must be > 0")
+
+    if min_img > max_img:
+        raise QuantumError("max_pic must be >= min_pic")
+
+    if min_img >= 2 ** nqubits or max_img > 2 ** nqubits:
+        raise QuantumError("min_pic must be < 2**nqubits and max_pic must be <= 2**nqubits")
+
+    import random
+    pics = random.randint(min_img, max_img)
+    integers = random.sample(range(2 ** nqubits), pics)
+    if q_type == "int":
+        items = [IntQubit(i) for i in integers]
+    else:
+        items = [Qubit(IntQubit(i)) for i in integers]
+
+    return OracleGate(nqubits, lambda qubits: qubits in items)
+
+
+def g_bbht_search(qstate, oracle):
+    """G-BBHT-Search
+
+    G-BBHT-Search is an algorithm based on Grover's algorithm,
+    but designed for an unknown oracle function that returns 1
+    for an unknown number of qubit states. Its based on this
+    paper: Boyer, M., Brassard, G., Hoyer, P.,  Tapp, A. (1998).
+    Tight bounds on quantum searching: Progress of Physics,
+    46(4-5), 493-505.
+
+    Parameters
+    ==========
+
+    qstate : Qubit
+        State the G-BBHT-Search should be applied on
+    oracle : OracleGate
+        The black box operator that flips the sign of the desired basis qubits.
+
+    Returns
+    =======
+
+    (Qubit, int) : (single Qubit that fullfilles oracle, numberof grover iterations applied)
+
+    Examples
+    ========
+
+    G-BBHT-Search for a orcale that returns 1 for a unkown number of statesbetween 5 - 15::
+
+        >>> from sympy.physics.quantum.grover import random_oracle, superposition_basis, g_bbht_search
+        >>> basis_states = superposition_basis(4)
+        >>> oracle = random_oracle(4, min_img=5, max_img=15, q_type="bin")
+        >>> x = g_bbht_search(basis_states, oracle)[0]
+        >>> oracle.search_function(x)
+        True
+
+    """
+    import random
+    import math
+    from sympy.physics.quantum.qubit import measure_all_oneshot
+
+    max_iterations = 1
+    factor_iterations = 6 / 5
+    count_grover_iterations = 0
+
+    while True:
+        for _ in range(random.choice(list(range(0, int(math.ceil(max_iterations)))))):
```

Comment:
It seems to me that it would not be necessary to import `math.ceil` as `int` will truncate floats so that `int(math.ceil(max_iterations))` is `int(max_iterations) + 1` (except when max_iterations is 1). Also `range(0, n)` can be written `range(n)`.

Reply:
It also looks to me that the outer range should exclude 0 and start from 1. Perhaps it would be possible to use `random.randrange`.

Reply:
It is my understanding that `random.randrange` would do the same as the combination of `random.choice`, `list` and `range` without the extra step of explicitly constructing the list of numbers over the range. Is there some reason why that could not be used?